File: frontend/src/lib/wrapped.ts
import { api } from "./api";
import type { MediaType } from "./media";

export type TopGenre = {
  name: string;
  count: number;
};

export type WrappedMedia = {
  tmdb_id: number;
  media_type: MediaType;
  title: string;
  poster_url: string | null;
};

export type WrappedResponse = {
  year: number;
  total_hours: number;
  total_movies: number;
  total_shows: number;
  total_episodes: number;
  top_genres: TopGenre[];
  top_show: WrappedMedia | null;
  top_show_episode_count: number | null;
  top_movie: WrappedMedia | null;
  previous_year_hours: number;
  hours_change_pct: number | null;
};

export async function getWrapped(year?: number): Promise<WrappedResponse> {
  const { data } = await api.get<WrappedResponse>("/wrapped", { params: year ? { year } : {} });
  return data;
}

export type PublicWrappedResponse = WrappedResponse & {
  username: string;
};

// Endpoint público — sem token de autenticação (a página /w/:token é
// acessível sem login). O interceptor de auth em lib/api.ts simplesmente
// não anexa Authorization quando não há sessão, então dá pra usar a
// mesma instância do axios sem risco.
export async function getPublicWrapped(token: string, year?: number): Promise<PublicWrappedResponse> {
  const { data } = await api.get<PublicWrappedResponse>(`/public/wrapped/${token}`, {
    params: year ? { year } : {},
  });
  return data;
}

File: backend/app/schemas/wrapped.py
from typing import Literal

from pydantic import BaseModel


class TopGenre(BaseModel):
    name: str
    count: int


class WrappedMedia(BaseModel):
    tmdb_id: int
    media_type: Literal["movie", "tv"]
    title: str
    poster_url: str | None


class WrappedResponse(BaseModel):
    year: int
    total_hours: float
    total_movies: int
    total_shows: int
    total_episodes: int
    top_genres: list[TopGenre]
    top_show: WrappedMedia | None
    top_show_episode_count: int | None
    top_movie: WrappedMedia | None
    previous_year_hours: float
    hours_change_pct: float | None


class PublicWrappedResponse(WrappedResponse):
    """Mesma coisa que WrappedResponse, mas exposta via link público sem
    login — inclui o username pra a página saber de quem é o Wrapped."""

    username: str
